### PR TITLE
[WIP] Making tests BigList pass

### PR DESCRIPTION
**Pull request by mikaelbr**
I was looking into this parser and saw there were some types lacking. I was first thinking of things like `2e2`, `2e-2` etc. But it also seems there are some other issues, like negative numbers, arrays without `,` separator such as `[2 2]`. So I added running of the `BigList` and started to fix some of the issues.

I thought it better to open this PR but as WIP to show that this is being worked on and to get feedback if I'm doing anything wrong.

What do you think, @jaredly?

### CODE REVIEW

**Comment by jaredly** [comment]
Very cool, thanks for contributing! I'm not sure what you mean by "arrays without separator" -- are you saying that `[2 2]` should be valid json?

**Comment by mikaelbr** [comment]
Ah, sorry. Poor wording on my part. I’m saying the opposite. It should’t be valid json syntax but with the current implementation it seems to be.

File: src/Json.re [hunk 1]
```diff
@@ -347,10 +357,7 @@ and parseArrayValue = (text, pos) => {
     }
   }
   | ']' => ([value], pos + 1)
-  | _ => {
-    let (rest, pos) = parseArrayValue(text, pos);
-    ([value, ...rest], pos)
-  }
+  | _ => fail(text, pos, "unexpected character")
```
Comment (jaredly) [inline src/Json.re]: oh, now I understand -- you're saying that the parser so far was too lax :) actually I'm totally fine with that (given that I'm parsing comments, which aren't technically valid json).

File: src/Json.re [hunk 2]
```diff
@@ -236,16 +236,26 @@ let parseString = (text, pos) => {
...
   let len = String.length(text);
-  while (i^ < len && Char.code('0') <= Char.code(text.[i^]) && Char.code(text.[i^]) <= Char.code('9')) {
-    i := i^ + 1;
-    /* print_endline(">" ++ string_of_int(pos) ++ " : " ++ string_of_int(i^)); */
-  };
-  let s = String.sub(text, pos, i^ - pos);
-  /* print_endline(s); */
-  (Number(float_of_string(s)), i^)
+  switch (text.[i^]) {
+  | '-' =>  {
```
[the remark this answers is not in the document]
Comment (mikaelbr) [reply src/Json.re]: Hehe. I saw that myself. I'm doing some work on this right this moment and I think this should be fixed locally. I'm pushing some changes soon. Just trying to get exponent numbers first

File: src/Json.re (hunk 1 above)
Comment (mikaelbr) [reply src/Json.re]: Yeah, I'm all for the more lax parsing, but I think things like `[1 2 3 4]` maybe is taking it too far, or?

File: src/Json.re (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaredly) [reply src/Json.re]: awesome, thanks!

File: src/Json.re (hunk 1 above)
Comment (jaredly) [reply src/Json.re]: yeah, that's fine :) as long as this doesn't break any of the current tests

File: src/Json.re (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikaelbr) [reply src/Json.re]: But doing all this number parsing adds up complexity-wise, so I think I need more eyes on this and serious code review.

File: src/Json.re (hunk 1 above)
Comment (mikaelbr) [reply src/Json.re]: All your existing tests will be good in this. But I've changed some in the BigList to be more lax.

File: src/Json.re [hunk 3]
```diff
@@ -236,16 +245,39 @@ let parseString = (text, pos) => {
...
+    [text.[0]]
+  } else {
+    [text.[0], ...toCharList(stringTail(text), maxDepth - 1)]
+  }
+};
+
+let continousDigits = (text, pos, len) => {
   let i = ref(pos);
-  let len = String.length(text);
   while (i^ < len && Char.code('0') <= Char.code(text.[i^]) && Char.code(text.[i^]) <= Char.code('9')) {
```
Comment (mikaelbr) [inline src/Json.re]: I was thinking too naively with exponent numbers. I think we need to extend this to check for `e` or `E` (followed by `+`, `-` or ``), but once and only once.

**Comment by mikaelbr** [comment]
What do you think about the scope of this, @jaredly? Should I continue on the work, or how would you want it?

File: src/Json.re [hunk 4]
```diff
@@ -255,9 +255,19 @@ let rec toCharList = (text, maxDepth) => {
...
 let continousDigits = (text, pos, len) => {
   let i = ref(pos);
-  while (i^ < len && Char.code('0') <= Char.code(text.[i^]) && Char.code(text.[i^]) <= Char.code('9')) {
+  let dec = ref(false);
+  let isNumber = n =>
+    Char.code('0') <= Char.code(text.[n]) && Char.code(text.[n]) <= Char.code('9');
+  let isDec = n =>
+    dec^ == false && Char.code('.') == Char.code(text.[n]);
+
+  while (i^ < len && (isNumber(i^) || isDec(i^))) {
```
Comment (mikaelbr) [inline src/Json.re]: I'm not sure if this is the best way forward and if it "scales" to exponent numbers and is expressive enough to handle all cases in a readable way. But I wanted to hear what you think

File: test/BigList.re [hunk 5]
```diff
@@ -291,8 +289,16 @@ let go = (ensure) => {
   roundtrip(" [] ");
 
   /* number */
+  roundtrip("[1.]");
+  roundtrip("[-2.]");
```
Comment (mikaelbr) [inline test/BigList.re]: Also wonder if this should be allowed as it is more relaxed parser?

File: src/Json.re [hunk 6]
```diff
@@ -236,16 +245,50 @@ let parseString = (text, pos) => {
...
+  }
+};
+
+let continousDigits = (text, pos, len) => {
   let i = ref(pos);
-  let len = String.length(text);
-  while (i^ < len && Char.code('0') <= Char.code(text.[i^]) && Char.code(text.[i^]) <= Char.code('9')) {
+  let dec = ref(false);
+  let isNumber = n =>
+    Char.code('0') <= Char.code(text.[n]) && Char.code(text.[n]) <= Char.code('9');
```
Comment (jaredly) [inline src/Json.re]: I think this would make a lot more sense as
```
switch (text.[n]) {
| '0'..'9' => true
| _ => false
}
```

File: src/Json.re [hunk 7]
```diff
@@ -236,16 +245,50 @@ let parseString = (text, pos) => {
...
+  let isDec = n =>
+    dec^ == false && Char.code('.') == Char.code(text.[n]);
```
Comment (jaredly) [inline src/Json.re]: Why the char.code here? can't you do `'.' == text.[n]`?

Comment (mikaelbr) [reply src/Json.re]: True. Remnants from earlier. Fixed.

File: src/Json.re (hunk 6 above)
Comment (mikaelbr) [reply src/Json.re]: Agreed. Fixed

**Comment by mikaelbr** [comment]
Yeah. This got out of hand quick. There are many different things to think about and I've rewritten it multiple times already, so I was thinking of a more 'make it work and refactor' approach. But I think it's smart to get more eyes on this.

**Comment by jaredly** [comment]
If this works in it's current form, I'm happy to merge it and then take a turn at refactoring :) thanks for all the work!

**Comment by mikaelbr** [comment]
It works in the sense that original tests are still green.

**Comment by mikaelbr** [comment]
Thanks. I'll keep an eye on the refactoring and see how it can be improved.

**Comment by mikaelbr** [comment]
Thinking about it actually, this might be more elegant using two passes: one to take all valid tokens of a float, and another to validate (check for only one `e`/`E`, `.` etc)

**Comment by jaredly** [comment]
ok, got all those tests passing :)

**Pull request merged**